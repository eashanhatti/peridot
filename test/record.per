let {
  'val Ty : OType =
    signature {
      T : OType,
      x : T
    };

  'val record : #Ty [ T = OType, x = T ] =
    structure {
      T = OType,
      x = T
    };
} in {
  OType
}
