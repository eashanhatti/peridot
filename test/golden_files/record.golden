((fromList [], fromList [], fromList []),
 Declare
   Obj
   (Rigid
      (RNameIntro
         (UserName "<PLACEHOLDER>") Obj (Id {unId = 0})))
   (Rigid (TypeType Obj))
   (Define
      (Rigid
         (RNameIntro
            (UserName "<PLACEHOLDER>") Obj (Id {unId = 0})))
      (RecType
         (fromList
            [(Field {unField = "T"}, Rigid (TypeType Obj)),
             (Field {unField = "f"},
              ObjFunType
                Explicit
                (LocalVar (Index {unIndex = 0}))
                (LocalVar (Index {unIndex = 1})))]))
      (Declare
         Obj
         (Rigid
            (RNameIntro
               (UserName "<PLACEHOLDER>") Obj (Id {unId = 1})))
         (GlobalVar
            (Rigid
               (RNameIntro (UserName "Ty") Obj (Id {unId = 0})))
            False)
         (Declare
            Obj
            (Rigid
               (RNameIntro
                  (UserName "<PLACEHOLDER>") Obj (Id {unId = 2})))
            (RecType
               (fromList
                  [(Field {unField = "T"},
                    Rigid
                      (SingType (Rigid (TypeType Obj)) (Rigid TwoType))),
                   (Field {unField = "f"},
                    ObjFunType
                      Explicit
                      (LocalVar (Index {unIndex = 0}))
                      (LocalVar (Index {unIndex = 1})))]))
            (Declare
               Obj
               (Rigid
                  (RNameIntro
                     (UserName "<PLACEHOLDER>") Obj (Id {unId = 3})))
               (RecType
                  (fromList
                     [(Field {unField = "T"},
                       Rigid
                         (SingType (Rigid (TypeType Obj)) (Rigid TwoType))),
                      (Field {unField = "f"},
                       ObjFunType
                         Explicit
                         (LocalVar (Index {unIndex = 0}))
                         (LocalVar (Index {unIndex = 1})))]))
               (Declare
                  Obj
                  (Rigid
                     (RNameIntro
                        (UserName "<PLACEHOLDER>") Obj (Id {unId = 4})))
                  (GlobalVar
                     (Rigid
                        (RNameIntro (UserName "Ty") Obj (Id {unId = 0})))
                     False)
                  (Define
                     (Rigid
                        (RNameIntro
                           (UserName "<PLACEHOLDER>") Obj (Id {unId = 1})))
                     (RecIntro
                        (fromList
                           [(Field {unField = "T"}, Rigid TwoType),
                            (Field {unField = "f"},
                             ObjFunIntro (LocalVar (Index {unIndex = 0})))]))
                     (Define
                        (Rigid
                           (RNameIntro
                              (UserName "<PLACEHOLDER>") Obj (Id {unId = 2})))
                        (RecIntro
                           (fromList
                              [(Field {unField = "T"},
                                Rigid (SingIntro (Rigid TwoType))),
                               (Field {unField = "f"},
                                ObjFunIntro (LocalVar (Index {unIndex = 0})))]))
                        (Define
                           (Rigid
                              (RNameIntro
                                 (UserName "<PLACEHOLDER>")
                                 Obj
                                 (Id {unId = 3})))
                           (RecIntro
                              (fromList
                                 [(Field {unField = "T"},
                                   Rigid
                                     (SingIntro
                                        (RecElim
                                           (GlobalVar
                                              (Rigid
                                                 (RNameIntro
                                                    (UserName "foo")
                                                    Obj
                                                    (Id {unId = 1})))
                                              False)
                                           (Field {unField = "T"})))),
                                  (Field {unField = "f"},
                                   RecElim
                                     (GlobalVar
                                        (Rigid
                                           (RNameIntro
                                              (UserName "foo")
                                              Obj
                                              (Id {unId = 1})))
                                        False)
                                     (Field {unField = "f"}))]))
                           (Define
                              (Rigid
                                 (RNameIntro
                                    (UserName "<PLACEHOLDER>")
                                    Obj
                                    (Id {unId = 4})))
                              (RecIntro
                                 (fromList
                                    [(Field {unField = "T"},
                                      SingElim
                                        (RecElim
                                           (GlobalVar
                                              (Rigid
                                                 (RNameIntro
                                                    (UserName "bar")
                                                    Obj
                                                    (Id {unId = 2})))
                                              False)
                                           (Field {unField = "T"}))),
                                     (Field {unField = "f"},
                                      RecElim
                                        (GlobalVar
                                           (Rigid
                                              (RNameIntro
                                                 (UserName "bar")
                                                 Obj
                                                 (Id {unId = 2})))
                                           False)
                                        (Field {unField = "f"}))]))
                              (Rigid (TypeType Obj))))))))))))
